designs/disruption-controls-by-action.md

```diff
@@ -0,0 +1,218 @@
+# Disruption Controls By Action 
+
+## Motivation
+Users want the ability to have different disruption controls for the budgets. The AKS Provider Drives NodeImageUpgrade, and K8sVersionUpgrade through Drift. Users may want to specify different disruption settings for upgrade, since upgrade is a fundementally different type of disruption in comparison to consolidation for example. Since upgrade changes the behavior of the cluster, its disruption schedule may be different from consolidation.
+
+## Desired Behavior + Requirements 
+1. User needs to be able to define an action and a budget/budgets for that action. 
+2. Actions that should be supported should by any disruption actions effected by the current Budgets implementation. (Consolidation, Emptiness, Expiration, Drift) 
+3. Budgets should still support the option to set a given behavior for all disruption actions. A disruption action `All` will represent behavior will be defined as a default for all actions. It should respect this budget for actions without an active disruption budget. 
+4. One should be able to specify for a given disruption action a NodeDisruptionBudget. If multiple budgets are active at a given time, karpenter will take action with the most restrictive budget
+## API Design
+### Approach A: Add an action field to disruption Budgets 
+This approach outlines a simple api change to the betav1 nodepool api. To allow disruption budgets to specify a disruption action. Thats it. 
+### Proposed Spec
+Add a simple field "action" is proposed to be added to the budgets. 
+```go
+// Budget defines when Karpenter will restrict the
+// number of Node Claims that can be terminating simultaneously.
+type Budget struct {
+	// Action defines the disruption action that this particular disruption budget applies to. 
+	Action DisruptionAction `json:"action" hash:"ignore"`
```

Comment:
Food for thought: I'm not sure what the right field name is here. I could go either way between `Action` and `Type`; though, I think that there is something to consider that "action" might imply to a user that this is the action that Karpenter is taking w.r.t. budget rather than the "action" that the budget is against.